bfvmm/include/intrinsics/x86/intel/apic/xapic.h

```diff
@@ -55,34 +166,242 @@ struct EXPORT_XAPIC xapic_control final : public lapic_control
+    {
+        auto offset = xapic::registers::tmr0.offset | idx;
+        return read_register(offset);
+    }
+
+    value_type read_irr(index idx) noexcept override
+    {
+        auto offset = xapic::registers::irr0.offset | idx;
+        return read_register(offset);
+    }
+
+    value_type read_lvt(lvt_reg reg) noexcept override
+    {
+        switch (reg) {
+            case cmci:
+                return read_register(xapic::registers::lvt_cmci.offset);
+            case timer:
+                return read_register(xapic::registers::lvt_timer.offset);
+            case thermal:
+                return read_register(xapic::registers::lvt_thermal.offset);
+            case perf:
+                return read_register(xapic::registers::lvt_perf.offset);
+            case lint0:
+                return read_register(xapic::registers::lvt_lint0.offset);
+            case lint1:
+                return read_register(xapic::registers::lvt_lint1.offset);
+            case error:
+                return read_register(xapic::registers::lvt_error.offset);
+
+            default:
+                bferror_info(0, "invalid lvt_reg");
+                return 0;
+        }
+    }
+
+    value_type read_count(count_reg reg) noexcept override
+    {
+        switch (reg) {
+            case initial:
+                return read_register(xapic::registers::init_count.offset);
+            case current:
+                return read_register(xapic::registers::cur_count.offset);
+
+            default:
+                bferror_info(0, "invalid count_reg");
+                return 0;
+        }
+    }
+
+    value_type read_div_config() noexcept override
+    { return read_register(xapic::registers::div_conf.offset); }
 
-    value_type read_lvt(lvt_reg reg) noexcept override;
-    value_type read_count(count_reg reg) noexcept override;
-    value_type read_div_config() noexcept override;
 
     //
     // Register writes
     //
-    void write_eoi() noexcept override;
-    void write_tpr(value_type tpr) noexcept override;
-    void write_svr(value_type svr) noexcept override;
-    void write_icr(value_type icr) noexcept override;
-    void write_lvt(lvt_reg reg, value_type val) noexcept override;
-    void write_init_count(value_type count) noexcept override;
-    void write_div_config(value_type config) noexcept override;
+    void write_eoi() noexcept override
+    { write_register(xapic::registers::eoi.offset, 0x0ULL); }
+
+    void write_tpr(value_type tpr) noexcept override
+    { write_register(xapic::registers::tpr.offset, tpr); }
+
+    void write_svr(value_type svr) noexcept override
+    { write_register(xapic::registers::sivr.offset, svr); }
+
+    void write_icr(value_type icr) noexcept override
+    {
+        value_type low = icr & 0x00000000FFFFFFFFULL;
+        value_type high = (icr & 0xFFFFFFFF00000000ULL) >> 32;
+        write_register(xapic::registers::icr_high.offset, high);
+        std::atomic_thread_fence(std::memory_order_release);
```

Comment:
We don't really need atomic here because this memory is local to the core. An sfence is all that is needed. Go ahead and create a fence intrinsic file with an sfence function. So the resulting code here would be
write...
intel_x64::sfence();
writre...